Update key repositories guide (#53)
denniskigen/o3-docs: pages/docs/key-repositories.en-US.mdx

# Key repositories

The code for O3 is hosted in a number of community-maintained GitHub repositories. O3 leverages a [microfrontends-based architecture](https://microfrontends.org) - an approach where the application is split into smaller, independent frontend applications that work together - meaning that the code for different parts of the application is spread across multiple repositories.

> These repositories are domain-specific, meaning that the code in each repository is typically focused on a particular concern. For example, the `Core` repository contains packages that handle cross-cutting concerns such as loading frontend modules, configuration, internationalization, rendering the UI, routing, error handling, styling, and so forth. The `Patient Chart` repository contains frontend modules that constitute widgets in a patient dashboard. Some of these repositories are structured as monorepos, meaning that they contain multiple packages, while others are more traditional, containing a single package. In cases where a repository contains multiple packages, it is usually because those packages share a single concern or theme. For example, the `Patient Chart` repository aggregates packages that handle concerns around a typical patient dashboard, including views for allergies, appointments, conditions, forms, medications, notes, programs, test results, vitals, and more.

Reviewer: ```suggestion
The code in each repository is focused on a particular concern. For example, the [Patient Chart](https://github.com/openmrs/openmrs-esm-patient-chart) repository contains frontend modules for the features of the patient chart, such as allergies, appointments, conditions, forms, medications, notes, programs, test results, and vitals. This is an example of a monorepo, where multiple packages are contained in a single repository; most O3 repositories are structured in this way. The [Core](https://github.com/openmrs/openmrs-esm-core) repository is special. It contains a few frontend modules, but also the core application logic that loads  frontend modules, as well as the [O3 tooling](https://github.com/openmrs/openmrs-esm-core/tree/main/packages/tooling/openmrs) and the [JavaScript API](https://github.com/openmrs/openmrs-esm-core/blob/main/packages/framework/esm-framework/docs/API.md).
```

Repositories you should know in O3 include:

> - [Core](#core) - aggregates the core frontend modules of the O3 frontend that handle essential cross-cutting concerns such as module loading, configuration, internationalization, UI rendering, feature flagging, routing, global state management, and more.

Reviewer: ```suggestion
- [Core](#core) - contains frontend modules for low-level concerns such as authentication. It also contains the base application which loads and stitches together all frontend modules, the app shell. This app shell provides the JavaScript library `@openmrs/esm-framework`, which provides for essential cross-cutting concerns such as configuration, internationalization, feature flagging, and global state management. This repository also contains the tooling for O3.
```

- [Cohort builder](#cohort-builder) - a tool that enables users to build and manage dynamic patient cohorts (groups) of patients based on multiple clinical and demographic criteria.
- [Module management](#module-management) - a learning frontend module based on the legacy OpenMRS Module management UI. The code in this repo is annotated with comments and can serve as a good starting point for new developers looking to understand how to build frontend modules.
- [Frontend RFC](#frontend-rfc) - a repository where architectural decisions affecting technologies and processes that apply to all new O3 distributions are first presented, argued, and finally agreed upon via consensus.
- [JSON schemas](#json-schemas) - a repository that hosts [standard JSON schemas](https://json-schema.org/) used in OpenMRS. These schemas are used to provide auto-completion and validation capabilities when editing routes registry files for O3 frontend modules and O3 form schemas.

### Core

> Core aggregates a mix of core apps that include both typical frontend modules as well as packages that don't render UIs. Typical frontend modules include:

Reviewer: ```suggestion
Core contains both typical frontend modules as well as packages that don't render UIs. Typical frontend modules include:
```

- [Primary navigation](https://github.com/openmrs/openmrs-esm-core/tree/main/packages/apps/esm-primary-navigation-app) - renders the top navigation menu in the application, featuring the compact Patient search widget, links to the Implementer tools menu, Registration page, User menu, and the App switcher.
- [Offline tools](https://github.com/openmrs/openmrs-esm-core/tree/main/packages/apps/esm-offline-tools-app) - responsible for rendering the widgets related to offline mode.

In addition to these UI-related packages, Core also includes the following non-UI packages:

> - [Core framework](https://github.com/openmrs/openmrs-esm-core/tree/main/packages/framework) - aggregates packages that make up the core O3 framework library, including shared API objects (such as the patient and visit objects), the breadcrumbs menu, the extension system, shared React utils, global state management tools, error handling, and the style guide, among other things. The functionality exported by these packages gets aggregated into an NPM JavaScript package called [esm-framework](http://npm.im/@openmrs/esm-framework) that gets reused by most frontend modules.

Reviewer: ```suggestion
- [Core framework](https://github.com/openmrs/openmrs-esm-core/tree/main/packages/framework) - aggregates packages that make up the core [O3 framework API](https://github.com/openmrs/openmrs-esm-core/blob/main/packages/framework/esm-framework/docs/API.md), including shared API objects (such as the patient and visit objects), the breadcrumbs menu, the extension system, shared React utils, global state management tools, error handling, and the style guide, among other things. The functionality exported by these packages gets aggregated into an NPM JavaScript package called [esm-framework](http://npm.im/@openmrs/esm-framework) that gets reused by most frontend modules.
```

### React form engine

> The [React form engine](https://github.com/openmrs/openmrs-form-engine-lib) is a React library that enables the creation and rendering of standardized medical forms within the O3 ecosystem. It provides a flexible, JSON-based form schema system with built-in validation, conditional rendering, and multilingual support - making it ideal for complex healthcare data collection needs. O3 form schemas are written in JSON and conform to the [O3 standard JSON schema spec](https://github.com/openmrs/openmrs-contrib-json-schemas/blob/main/form.schema.json) which defines the structure and constraints of the form data. The [O3 reference application](https://o3.openmrs.org/openmrs/spa) ships an embedded [Form Builder](https://github.com/openmrs/openmrs-esm-form-builder) that users can leverage to build forms interactively, using the interactive schema builder, or by writing code into the built-in JSON schema editor. Learn more about the React form engine from the [official docs](https://openmrs.atlassian.net/wiki/spaces/projects/pages/68747273/O3+Form+Docs).

Reviewer: ```suggestion
The [React form engine](https://github.com/openmrs/openmrs-form-engine-lib) is a library that enables creating and rendering medical forms within O3. It provides a flexible, JSON-based form schema with built-in validation, conditional rendering, and multilingual support. O3 form schemas conform to the [O3 standard JSON schema spec](https://github.com/openmrs/openmrs-contrib-json-schemas/blob/main/form.schema.json) which defines the structure and constraints of the form data.   Learn more about the React form engine from the [official docs](https://openmrs.atlassian.net/wiki/spaces/projects/pages/68747273/O3+Form+Docs). The [O3 reference application](https://o3.openmrs.org/openmrs/spa) ships an embedded [Form Builder](https://github.com/openmrs/openmrs-esm-form-builder) to make building forms easy.
```

...and then the reader reads on to the next section.

BTW, my preference would be to capitalize things like React Form Engine and Form Builder as proper nouns.

### Form builder

> The [Form builder](https://github.com/openmrs/openmrs-esm-form-builder) is a frontend module used to create OpenMRS form schemas. It enables users to both create new schemas and edit existing ones. It provides an embedded code editor that accepts JSON code and an interactive editor where users can construct a schema interactively without writing code. Schemas built using the form builder can be persisted to the server, and optionally published, which makes them available to the frontend via the forms widget. Schemas built in the form builder get rendered via the React form engine, which allows users to validate and test their forms straight from the form builder UI.

Reviewer: ```suggestion
The [Form builder](https://github.com/openmrs/openmrs-esm-form-builder) is a frontend module used to create OpenMRS form schemas, for use with the React form engine. It enables users to both create new schemas and edit existing ones. It provides an embedded code editor that facilitates writing JSON code, and a WYSIWYG schema builder interface where users can construct a schema interactively without writing code. Schemas built using the form builder can be persisted to the server, and optionally published, which makes them available to the frontend via the forms widget. The React form engine is used in the form builder to validate and test forms directly in the form builder UI.
```

### Fast data entry

> The [Fast data entry app](https://github.com/openmrs/openmrs-esm-fast-data-entry-app) is a frontend module that allows for a natural workflow when entering many pre-recorded forms at a time. It is not meant for point-of-care workflows, but rather as a way to do retrospective data entry. It allows the rapid input of forms for group sessions enabling recording information about a group session or for session participants.

Reviewer: ```suggestion
The [Fast data entry app](https://github.com/openmrs/openmrs-esm-fast-data-entry-app) is a frontend module that provides a natural workflow for entering many pre-recorded forms at a time. It is not meant for point-of-care workflows, but rather as a way to do retrospective data entry. It also can be used for the rapid input of forms for group sessions.
```

### Medication dispensing

> The [Medication dispensing app](https://github.com/openmrs/openmrs-esm-dispensing-app) is a simple tool that enables tracking dispensed medications. It's designed to serve small sites that don't want to deal with implementing a whole other external supply chain/ERP system. These sites typically only care about tracking what needs to be dispensed vs what has already been dispensed. The primary target users for this functionality are implementations and facilities that have an on-site pharmacy and don't charge the patient/customer for the services, who want to dispense medications based on orders within the same OpenMRS instance. Billing and inventory are out of scope.

Reviewer: ```suggestion
The [Medication dispensing app](https://github.com/openmrs/openmrs-esm-dispensing-app) is a simple tool for tracking dispensed medications. It's designed to serve small sites that don't want to deal with implementing a whole external supply chain/ERP system. It simply tracks what needs to be dispensed and what has already been dispensed. The primary target users for this functionality are implementations and facilities that have an on-site pharmacy and don't charge the patient/customer for the services, and who want to dispense medications based on orders within the same OpenMRS instance. Billing and inventory are out of scope.
```
